agent/src/discovery_handler_manager/discovery_handler_registry.rs

```diff
@@ -0,0 +1,979 @@
+
+    /// Register a new endpoint to make it available to all current and future queries
+    async fn register_endpoint(&self, endpoint: Arc<dyn DiscoveryHandlerEndpoint>);
+}
+
+/// Real world implementation of the Discovery Handler Request
+struct DHRequestImpl {
+    endpoints: RwLock<Vec<watch::Receiver<Vec<Arc<DiscoveredDevice>>>>>,
+    notifier: watch::Sender<Vec<Arc<DiscoveredDevice>>>,
+    key: String,
+    handler_name: String,
+    details: String,
+    properties: Vec<DiscoveryProperty>,
+    extra_device_properties: HashMap<String, String>,
+    kube_client: Arc<dyn DiscoveryManagerKubeInterface>,
+    termination_notifier: Arc<Notify>,
+}
+
+impl DiscoveryHandlerRequest for DHRequestImpl {
+    fn get_instances(&self) -> Result<Vec<Instance>, DiscoveryError> {
+        Ok(self
+            .notifier
+            .borrow()
+            .iter()
+            .map(|i| self.device_to_instance(i))
+            .collect())
+    }
+}
+
+impl DHRequestImpl {
+    fn device_to_instance(&self, dev: &DiscoveredDevice) -> Instance {
+        let (rdev, shared) = match dev {
+            DiscoveredDevice::LocalDevice(d, _) => (d, false),
+            DiscoveredDevice::SharedDevice(d) => (d, true),
+        };
+        let mut properties = rdev.properties.clone();
+        properties.extend(self.extra_device_properties.clone());
+        Instance {
+            spec: InstanceSpec {
+                cdi_name: self.get_device_cdi_fqdn(dev),
+                configuration_name: self.key.clone(),
+                broker_properties: properties,
+                shared,
+                nodes: Default::default(),
+                device_usage: Default::default(),
+                capacity: Default::default(),
+            },
+            metadata: ObjectMeta {
+                name: Some(format!("{}-{}", self.key, dev.device_hash())),
+                ..Default::default()
+            },
+        }
+    }
+
+    fn get_device_cdi_fqdn(&self, dev: &DiscoveredDevice) -> String {
+        format!("akri.sh/{}={}", self.key, dev.device_hash())
+    }
+
+    async fn watch_devices(&self, mut rec: broadcast::Receiver<Arc<dyn DiscoveryHandlerEndpoint>>) {
+        loop {
+            let mut local_endpoints = self.endpoints.write().await.clone();
+            let futures = local_endpoints.iter_mut().map(|e| e.changed().boxed());
+            select! {
+                (a, index, _) = select_all(futures) => {
+                    if a.is_err() {
+                        let mut write_endpoint = self.endpoints.write().await;
+                        write_endpoint.remove(index);
+                        if write_endpoint.is_empty() {
+                            return;
+                        }
+                    }
+                },
+                Ok(endpoint) = rec.recv() => {
+                    if endpoint.get_name() != self.handler_name {
+                        // We woke up for another kind of DH, let's get back to sleep
```

Comment:
Will rename `rec` and `endpoint` here to be more explicit, this can happen if another discovery handler registers itself:
We are awaking all active requests whenever a new discovery handler registers itself, so that if it is a DH with the same name/kind as the one of this request we also send the request to that newly registered endpoint.
To simplify I chose to have a single MQ to broadcast those registrations (that should be pretty rare anyway), rather than creating one per name/kind of DH.